src/NUnitFramework/framework/Internal/TestExecutionContext.cs

```diff
@@ -493,6 +494,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
             CurrentValueFormatter = formatterFactory(CurrentValueFormatter);
         }
 
+        /// <summary>
+        /// Attach a file to the current test result
+        /// </summary>
+        /// <param name="filepath">Filepath to attachment. Can be absolute, or relative to working directory.</param>
+        /// <param name="description">Optional description of attachment</param>
+        public void AddTestAttachment(string filepath, string description = null)
+        {
+            Guard.ArgumentNotNull(filepath, nameof(filepath));
+            Guard.ArgumentValid(filepath.IndexOfAny(Path.GetInvalidPathChars()) == -1, 
+                $"Test attachment filepath contains invalid path characters. {filepath}", nameof(filepath));
+
+            if (!Path.IsPathRooted(filepath))
+                filepath = Path.Combine(TestContext.CurrentContext.WorkDirectory, filepath);
+
+#if !PORTABLE
+            if (!File.Exists(filepath))
+                throw new FileNotFoundException("Test attachment filepath could not be found.", filepath);
```

Comment:
Typically, if you're going to do anything else with the file other than verify that it exists, the existence check is meaningless because it's race condition prone. To have meaningful handling you'd have to catch FileNotFoundException because it's the only atomic API.

> I imagine a common use case may be where people only refer to attachments when the test has failed - so broken paths may only be discovered 6 months after writing, when the test first fails, and it's too late.

But this is a great point and would be very helpful.